src/edu/csus/ecs/pc2/core/report/ResultsCompareReport.java

```diff
@@ -0,0 +1,369 @@
+// Copyright (C) 1989-2023 PC2 Development Team: John Clevenger, Douglas Lane, Samir Ashoo, and Troy Boudreau.
+package edu.csus.ecs.pc2.core.report;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import edu.csus.ecs.pc2.VersionInfo;
+import edu.csus.ecs.pc2.core.Constants;
+import edu.csus.ecs.pc2.core.IInternalController;
+import edu.csus.ecs.pc2.core.Utilities;
+import edu.csus.ecs.pc2.core.imports.clics.FieldCompareRecord;
+import edu.csus.ecs.pc2.core.imports.clics.FileComparison;
+import edu.csus.ecs.pc2.core.log.Log;
+import edu.csus.ecs.pc2.core.model.ClientSettings;
+import edu.csus.ecs.pc2.core.model.Filter;
+import edu.csus.ecs.pc2.core.model.IInternalContest;
+import edu.csus.ecs.pc2.core.report.FileComparisonUtilities.AwardKey;
+import edu.csus.ecs.pc2.core.report.FileComparisonUtilities.ResultTSVKey;
+import edu.csus.ecs.pc2.core.report.FileComparisonUtilities.ScoreboardJSONKey;
+import edu.csus.ecs.pc2.exports.ccs.ResultsFile;
+
+/**
+ * Compare contest results files report.
+ * 
+ * @author Douglas A. Lane <pc2@ecs.csus.edu>
+ */
+public class ResultsCompareReport implements IReport {
+
+    private String primaryCCSResultsDir = null;
+
+    private String pc2ResultsDir = null;
+
+    private IInternalContest contest;
+
+    private IInternalController controller;
+
+    private Log log;
+
+    private Filter filter = new Filter();
+
+    private boolean fullDetails = false;
+    
+    private String missingSourceMessage = ComparisonState.MISSING_SOURCE.toString(); 
+    
+    private String missingTargetMessage = ComparisonState.MISSING_TARGET.toString();
+
+    /**
+     * 
+     */
+    private static final long serialVersionUID = -796328654541676730L;
+
+    public ResultsCompareReport(IInternalContest contest, IInternalController controller, String primaryCCSResultsDir, String pc2ResultsDir, boolean fullDetails) {
+        this(contest, controller, primaryCCSResultsDir, pc2ResultsDir, fullDetails, null, null);
+    }
+    
+    public ResultsCompareReport(IInternalContest contest, IInternalController controller, String primaryCCSResultsDir, String pc2ResultsDir, boolean fullDetails, String missingSourceMessage,
+            String missingTargetMessage) {
+        super();
+        this.primaryCCSResultsDir = primaryCCSResultsDir;
+        this.pc2ResultsDir = pc2ResultsDir;
+        this.fullDetails = fullDetails;
+        if (missingTargetMessage != null) {
+            this.missingTargetMessage = missingTargetMessage;
+        }
+        if (missingSourceMessage != null) {
+            this.missingSourceMessage = missingSourceMessage;
+        }
+        setContestAndController(contest, controller);
+    }
+    
+
+    public ResultsCompareReport() {
```

Comment:
I think this constructor is a bad idea.  It allows someone to construct a `ResultsCompareReport` which has no `contest` or `controller` -- potentially resulting in `NullPointerExceptions`.  I think the constructor should be removed, unless a good explanation for it exists (in which case such explanation should be documented in a header comment).